util/string_ref.c

```diff
@@ -0,0 +1,125 @@
+/*
+   Copyright 2020 Bloomberg Finance L.P.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+
+#include <plhash.h>
+#include <assert.h>
+#include <pthread.h>
+#include "string_ref.h"
+#include "locks_wrap.h"
+#include "comdb2_atomic.h"
+#include "logmsg.h"
+
+#define TRACK_REFERENCES
+
+#ifdef TRACK_REFERENCES
+static hash_t *hrefs = NULL;
+static pthread_mutex_t hmtx = PTHREAD_MUTEX_INITIALIZER;
+#endif
+int gbl_creation_count;
+
+struct string_ref_t {
+    char *str;
+    int cnt;
+};
+
+
+/* Makes a copy of the string passed and uses that as a reference counted object
+ */
+string_ref_t * create_string_ref(const char *str)
+{
+    assert(str);
+    int len = strlen(str);
+    string_ref_t *ref = malloc(sizeof(string_ref_t) + len + 1);
+    ref->cnt = 1;
+    ref->str = ((char*)ref) + sizeof(string_ref_t);
+    strcpy(ref->str, str);
+
+#ifdef TRACK_REFERENCES
+    Pthread_mutex_lock(&hmtx);
+    if(!hrefs) 
+        hrefs = hash_init(sizeof(&ref));
+    hash_add(hrefs, ref);
+    gbl_creation_count += 1;
+    Pthread_mutex_unlock(&hmtx);
+#else
+    ATOMIC_ADD32(gbl_creation_count, 1);
+#endif
+    return ref;
+}
+
+/* Get a reference by increasing the count */
+string_ref_t * get_ref(string_ref_t *ref)
+{
+    assert(ref);
+    int cnt = ATOMIC_ADD32(ref->cnt, 1);
```

Comment:
Do you intend to make the function thread-safe? If so, you need a mutex to protect the whole block here. If not, just drop `ATOMIC_ADD32`.

Reply:
So, you can call `get_ref()` as many times as you need--as long as you `put_ref()` when you are done, we will eventually free the string. In order to do that accounting correctly, we want to use atomic_add. Notice that if from two threads, one calls get_ref() and the other calls put_ref(), we should get correct cnt via the atomics. Notice also that you can never get to cnt == 0 unless that was your truly your last reference that you are calling put_ref() on--after you put_ref the object will be set to NULL. Now of course one can cheat and make a copy of the str_ref, but that would be a logical error for this type of object. Unfortunately we don't have the luxury of disabling assignment operator in the c++ style.

Reply:
Yes, I misread. The local variable `cnt` is used below instead of `ref->cnt`, so it's fine.